### PR TITLE
feat(bundles): relate bundle events together

### PR DESCRIPTION
**Pull request by james-a-morris**
This PR adds out-of-band associations between proposed root bundle events, disputed root bundle events, and canceled root bundle events.

### CODE REVIEW

File: docker-compose.yml
Comment (james-a-morris) [inline docker-compose.yml:1]: We can override the ENV of `postgres` and `redis` so it points to the internal service

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  relayerRefundRoot: string;
+
+  @Column()
+  slowRelayRoot: string;
+
+  @OneToOne(() => ProposedRootBundle, { nullable: false })
+  @JoinColumn({
+    foreignKeyConstraintName: "FK_bundle_rootBundleProposeId",
+  })
+  proposal: ProposedRootBundle;
```
Comment (amateima) [inline packages/indexer-database/src/entities/bundle/Bundle.ts:41]: For all the events, can you add an `<event_name>Id` column? These columns are already created by the migration files and would be nice to have access to the id without performing the join.

File: packages/indexer/src/services/bundles.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+      message: "Postgres connection not provided",
+    });
+    throw new IndexerConfigurationMalformedError();
+  }
+
+  return async () => {
+    await Promise.all([
+      assignBundleToProposedEvent(postgres, logger),
+      asignDisputeEventToBundle(postgres, logger),
+      asignCanceledEventToBundle(postgres, logger),
```
Comment (amateima) [inline packages/indexer/src/services/bundles.ts:47]: Shouldn't these function be executed serially as they all depend on the Bundle entity to be create first?

Comment (james-a-morris) [reply packages/indexer/src/services/bundles.ts:47]: They don't need to be given that each one is isolated and accounts for the issue that no bundle event exists.

File: packages/indexer-database/src/entities/index.ts [hunk 3]
```diff
@@ -11,5 +11,8 @@ export * from "./evm/RequestedSpeedUpV3Deposit";
 export * from "./evm/RelayedRootBundle";
 export * from "./evm/ExecutedRelayerRefundRoot";
 export * from "./evm/TokensBridged";
+
+// Bundle Metaclass
+export * from "./bundle/Bundle";
```
Comment (melisaguevara) [inline packages/indexer-database/src/entities/index.ts:16]: I was thinking on the folders inside /entities to be the different schemas. wdyt about keeping this directly in the /entities folder?

File: packages/indexer/src/main.ts [hunk 4]
```diff
@@ -154,10 +154,17 @@ export async function Main(
...
+    logger,
+    redis,
+    postgres,
+  });
+
   // TODO: add looping to keep process going
   // do {
   logger.info("index loop starting");
   await depositIndexer(Date.now());
+  await bundleIndexer();
```
Comment (melisaguevara) [inline packages/indexer/src/main.ts:167]: I associate `Indexer` with the indexing of the raw events. IMO we should name this different, maybe `bundleProcessor` or something like that? wdyt?

File: packages/indexer-database/src/migrations/1725910210791-Bundle.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+export class Bundle1725910210791 implements MigrationInterface {
+  name = "Bundle1725910210791";
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TYPE "aggregate"."bundle_status_enum" AS ENUM('Proposed', 'Validated', 'Canceled', 'Disputed', 'Executed')`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "aggregate"."bundle" ("id" SERIAL NOT NULL, "poolRebalanceRoot" character varying NOT NULL, "relayerRefundRoot" character varying NOT NULL, "slowRelayRoot" character varying NOT NULL, "status" "aggregate"."bundle_status_enum" NOT NULL, "proposalId" integer, "cancellationId" integer, "executionId" integer, "disputeId" integer, CONSTRAINT "REL_a8344aa79161a63b6397cc8006" UNIQUE ("proposalId"), CONSTRAINT "REL_0b9be31e1e95d2f01f834b6509" UNIQUE ("cancellationId"), CONSTRAINT "REL_830aeee2e34ac04c5d5a604046" UNIQUE ("executionId"), CONSTRAINT "REL_707430c410bc8a69af9432bedf" UNIQUE ("disputeId"), CONSTRAINT "PK_637e3f87e837d6532109c198dea" PRIMARY KEY ("id"))`,
+    );
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/migrations/1725910210791-Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/647e950b327e3d9f65ffe1587ff5f9dce7485e61

File: packages/indexer-database/src/migrations/1725998563558-Bundle.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
+import { MigrationInterface, QueryRunner } from "typeorm";
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/migrations/1725998563558-Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/647e950b327e3d9f65ffe1587ff5f9dce7485e61

File: packages/indexer-database/src/entities/bundle/Bundle.ts (hunk 1 above)
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts:41]: https://github.com/across-protocol/indexer/pull/25/commits/12c0ec7a2e176f412233c9a959dbd904aaff15e4

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+  @ManyToMany(() => RootBundleExecuted)
+  @JoinTable({
+    name: "bundle_executions_join", // Custom join table name
+    joinColumn: {
+      name: "bundle_id",
+      referencedColumnName: "id",
+    },
+    inverseJoinColumn: {
+      name: "execution_id",
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/12c0ec7a2e176f412233c9a959dbd904aaff15e4

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  @JoinColumn({
+    foreignKeyConstraintName: "FK_bundle_rootBundleCanceledId",
+  })
+  cancelation: RootBundleCanceled;
+
+  @ManyToMany(() => RootBundleExecuted)
+  @JoinTable({
+    name: "bundle_executions_join", // Custom join table name
+    joinColumn: {
+      name: "bundle_id",
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/12c0ec7a2e176f412233c9a959dbd904aaff15e4

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+  @OneToOne(() => RootBundleCanceled, { nullable: true })
+  @JoinColumn({
+    foreignKeyConstraintName: "FK_bundle_rootBundleCanceledId",
+  })
+  cancelation: RootBundleCanceled;
+
+  @ManyToMany(() => RootBundleExecuted)
+  @JoinTable({
+    name: "bundle_executions_join", // Custom join table name
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/12c0ec7a2e176f412233c9a959dbd904aaff15e4

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+export enum BundleStatus {
+  Proposed = "Proposed",
+  Validated = "Validated",
+  Canceled = "Canceled",
+  Disputed = "Disputed",
+  Executed = "Executed",
+}
+
+@Entity({ schema: "aggregate" })
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/12c0ec7a2e176f412233c9a959dbd904aaff15e4

File: packages/indexer/src/services/bundles.ts (hunk 2 above)
Comment (james-a-morris) [reply packages/indexer/src/services/bundles.ts]: https://github.com/across-protocol/indexer/pull/25/commits/a29450d5608cc28ffe87e3f9cb8ce5a524d3e831

File: packages/indexer/src/services/bundles.ts [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+      at: "Bundles#Indexer",
+      message: "Postgres connection not provided",
+    });
+    throw new IndexerConfigurationMalformedError();
+  }
+
+  return async () => {
+    await Promise.all([
+      assignBundleToProposedEvent(postgres, logger),
+      asignDisputeEventToBundle(postgres, logger),
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer/src/services/bundles.ts]: https://github.com/across-protocol/indexer/pull/25/commits/a29450d5608cc28ffe87e3f9cb8ce5a524d3e831

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  })
+  proposal: ProposedRootBundle;
+
+  @OneToOne(() => RootBundleCanceled, { nullable: true })
+  @JoinColumn({
+    foreignKeyConstraintName: "FK_bundle_rootBundleCanceledId",
+  })
+  cancelation: RootBundleCanceled;
+
+  @ManyToMany(() => RootBundleExecuted)
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts:67]: Correct - we could use a one->many if we were able to add an FK constraint to the executions

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  PrimaryGeneratedColumn,
+} from "typeorm";
+import { ProposedRootBundle } from "../evm/ProposedRootBundle";
+import { RootBundleCanceled } from "../evm/RootBundleCanceled";
+import { RootBundleExecuted } from "../evm/RootBundleExecuted";
+import { RootBundleDisputed } from "../evm/RootBundleDisputed";
+
+export enum BundleStatus {
+  Proposed = "Proposed",
+  Validated = "Validated",
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts:17]: The idea would be to map the case between executed and proposed when a bundle has passed liveness but hasn't been executed.

File: packages/indexer/src/main.ts (hunk 4 above)
Comment (james-a-morris) [reply packages/indexer/src/main.ts]: Good call - replaced the namings: https://github.com/across-protocol/indexer/pull/25/commits/8d2a2204258cf749482ff632e497ac244f8e1fcc

File: packages/indexer-database/src/entities/index.ts (hunk 3 above)
Comment (james-a-morris) [reply packages/indexer-database/src/entities/index.ts]: Done: https://github.com/across-protocol/indexer/pull/25/commits/8d2a2204258cf749482ff632e497ac244f8e1fcc

File: packages/indexer/src/main.ts [hunk 14]
```diff
@@ -163,9 +163,19 @@ export async function Main(
...
+  ]);
+
+  logger.info({
+    at: "Indexer#Main",
+    message: "Indexer loop completed",
+    results: {
+      depositIndexerRunSuccess: depositResults.status === "fulfilled",
+      bundleProcessorRunSuccess: bundleResults.status === "fulfilled",
+    },
+  });
```
Comment (james-a-morris) [inline packages/indexer/src/main.ts:178]: Added a bit more context to the results of these runs.

File: packages/indexer-database/src/entities/bundle/Bundle.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+  @OneToOne(() => RootBundleDisputed, { nullable: true })
+  @JoinColumn({
+    foreignKeyConstraintName: "FK_bundle_rootBundleDisputedId",
+  })
+  dispute: RootBundleDisputed;
+
+  @Column({ type: "enum", enum: BundleStatus })
+  status: BundleStatus;
+}
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/bundle/Bundle.ts]: Done: https://github.com/across-protocol/indexer/pull/25/commits/6eddc931760f555008a593cb6e218713c9433b9e

File: packages/indexer/src/services/bundles.ts [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+  const canceledRootBundleRepository = dataSource.getRepository(
+    entities.RootBundleCanceled,
+  );
+  // Find all canceled events that haven't been associated with a bundle.
+  const canceledEventsWithoutBundleAssociated =
+    await canceledRootBundleRepository
+      .createQueryBuilder("drb")
+      .select(["drb.id", "drb.blockNumber", "drb.logIndex"])
+      .leftJoin("bundle", "b", "b.cancellationId = drb.id")
+      .where("b.cancellationId IS NULL")
```
Comment (melisaguevara) [inline packages/indexer/src/services/bundles.ts:165]: This field is named as `cancelationId` in the entity

Comment (james-a-morris) [reply packages/indexer/src/services/bundles.ts]: https://github.com/across-protocol/indexer/pull/25/commits/af66a2e0280059c4db0d4ae4531fff9b5fc3caa7

File: packages/indexer-database/src/migrations/1726082056015-Bundle.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    await queryRunner.query(
+      `CREATE TYPE "public"."bundle_status_enum" AS ENUM('Proposed', 'Validated', 'Canceled', 'Disputed', 'Executed')`,
+    );
+    await queryRunner.query(
+      `CREATE TABLE "bundle" (
+          "id" SERIAL NOT NULL, 
+          "poolRebalanceRoot" character varying NOT NULL, 
+          "relayerRefundRoot" character varying NOT NULL, 
+          "slowRelayRoot" character varying NOT NULL, 
+          "proposalnId" integer,
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/migrations/1726082056015-Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/c4d7df93ce208c38339d4b7232777533b7416eae

File: packages/indexer-database/src/entities/Bundle.ts [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+  @Column()
+  slowRelayRoot: string;
+
+  @OneToOne(() => ProposedRootBundle, { nullable: false })
+  @JoinColumn({
+    foreignKeyConstraintName: "FK_bundle_rootBundleProposeId",
+  })
+  proposal: ProposedRootBundle;
+
+  @Column({ nullable: true })
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/c4d7df93ce208c38339d4b7232777533b7416eae

File: packages/indexer-database/src/entities/Bundle.ts [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+  proposalnId: number;
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply packages/indexer-database/src/entities/Bundle.ts]: https://github.com/across-protocol/indexer/pull/25/commits/c4d7df93ce208c38339d4b7232777533b7416eae

**Review by melisaguevara — APPROVED** [review]
lgtm!

File: packages/indexer/src/main.ts (hunk 4 above)
Comment (amateima) [reply packages/indexer/src/main.ts]: Good call @melisaguevara

**Review by amateima — APPROVED** [review]

**Pull request merged**